### ISSUE TITLE
Entity ID Generator Sequence

### ISSUE THREAD

**Issue by dchiarad** [oasp/oasp4j#317]
I have created a new project by using the OASP template. My underlying database is MySQL which does not support database sequences. The problem is AbstractPersistenceEntity forces you to use the sequence generator on your entities even though the database does not support this kind of generator. In the end I had to cut the inheritance of my entities so that I can define the id and generator type in my ApplicationPersistenceEntity.

However, from my point of view there should be more freedom on how to created the primary id for entities.

**Comment by hohwille** [comment]
Thanks for your feedback. I totally see your point.
We are using getter annotation instead of private field annotation exactly for such reason.
Did you try to override getId() in ```ApplicationPersistenceEntity``` and change the behaviour there?

See also #347.

**Comment by hohwille** [comment]
BTW: In general the app template has to work out of the box for one default database. Shipping to another database such as MySQL (what is IMHO not the best choice compared to PostgreSQL) will require some changes. The issue here there should be an easier way to do that change and not to find a way that nothing has to be changed.

In PR #347 we should ensure that code such as DOAs still do not rely on abstract implementation.
Further we should have some documentation guide for using different databases. As we are moving to spring-boot anyway this guide will soon make sense and we can also link to the spring-boot guide for the configuration of arbitrary databases and data-sources.

**Comment by ivanderk** [comment]
Unfortunately, I don´t think it is possible to "override" the PK definition in the base class. I´ve tried it in several ways and it resulted in different errors. So we choose the same solution as @dchiarad and not inherit AbstractPersistenceEntity in order to support MySql. I would argue that best practice determines that the PK should not be defined in a framework-level base class.

MySql may not be ideal, far from it, but it is a [db engine in wide use](http://db-engines.com/en/ranking)
So we cannot be ignoring it.

**Comment by ivanderk** [comment]
Easiest way to address this would be to change in AbstractPersistenceEntity the @GeneratedValue strategy from GenerationType.SEQUENCE to  GenerationType.AUTO

This seems to work for MySQL. I don´t currently have the means to test this for Postgress, Oracle etc.

Any thoughts or comments before I would send in a Pull Request?

**Comment by hohwille** [comment]
We just created the new oasp4j-module-basic:
https://github.com/oasp/oasp4j/tree/develop/oasp4j-modules/oasp4j-basic

and esp. added this:
https://github.com/oasp/oasp4j/blob/develop/oasp4j-modules/oasp4j-jpa/src/main/java/io/oasp/module/jpa/dataaccess/api/AbstractPersistenceEntity.java

So what would be the best solution? 
* Adapt the build process that we also provide oasp4j-jpa-mysql where that class differs?
* Or should we investigate why hibernate does not allow the overriding?
* Or is there a solution possible that is working for all DBs?
* Or should we move all such code to the sample and application-template and not provide it as module?

**Comment by hohwille** [comment]
AUTO is IMHO the default. All we need to do is remove the strategy from @GeneratedValue and close this ticket.

**Comment by ivanderk** [comment]
I`ve been testing replacing @GeneratedValue(strategy = GenerationType.SEQUENCE) to @GeneratedValue(strategy = GenerationType.AUTO)
On the getID method in io.oasp.module.jpa.dataaccess.api.AbstractPersistenceEntity as in

  @Override
  @Id
  @GeneratedValue(strategy = GenerationType.AUTO)
  public Long getId() {

    return this.id;
  }

Now MYSQL works fine, as does MSQL : And now the batch unit test fail (ProductImportJobTest)  :-D
Reason : the exeception it generates complains about the fact that the Id == null when inserting the new data. Probaly because in case of AUTO H2 will use an IDENTIY generator and is seems that Hibernate will diable « the JDBC batch support for entities using the IDENTITY generator.» 
http://vladmihalcea.com/2014/07/08/hibernate-identity-sequence-and-table-sequence-generator/

Any ideas ?

**Comment by dumbNickname** [comment]
Just an idea - if you really want to be fully flexible then probably you would need to build something like a sequence generator provider, which would hide DB specific sequence settings and could be implemented individually for specific data bases. it is just a rough idea but I typed in google: "hibernate custom sequence" and e.g. this popped out: http://stackoverflow.com/questions/11631800/hibernate-how-specify-custom-sequence-generator-class-name-using-annotations

If i understand properly, OASP4J has one abstract entity, so its children would need a possibility to redefine the generator being used for them (did not check how inheritance works for those annotations). Then it would be even possible to reuse the same sequence generator across different data bases (might be useful when somebody keeps most of the data in relational DB and for some reason binary data or some documents in MongoDB, but would like to have consistent IDs). 

If this is a shitty approach after all, please ignore :P

**Comment by maihacke** [comment]
I think we should have closer look why batch breaks with jdbc batching disabled. If I remeber correctly there are some more problems with jdbc batching, eg when using blobs. So we should expect that this feature has to be disabled anyway. So batches should work without jdbc batching and obviously identiy columns should be supported with our batches.

**Comment by hohwille** [comment]
PR had been merged. Strategy is now `auto`
See:
https://github.com/oasp/oasp4j/blob/develop/modules/jpa/src/main/java/io/oasp/module/jpa/dataaccess/api/AbstractPersistenceEntity.java#L42

So I will close this as it is done.

**Comment by hohwille** [comment]
I was not entirely aware of the impact of this.
* Before we were using a sequence what is a best practice in projects that use reasonable databases that do support sequences.
* Because there are some crazy DBs out there like MySQL that do not support sequences we had a problem to support them with the pre-defined base-class.
* To be able to also support MySQL we now switched to `GenerationType.AUTO`.
* However, this breaks compatibility for DBs like Oracle.
* Even worse it might also be totally wrong to force people using Autoincrement. In my project we are following the best practice to define an ID-Range for master-data and test-data. We initialize the hibernate sequence with a very high value above that ranges so we have room for upcoming master-data that we want to ship with flyway in upcoming releases in the next 10 years. This is now completely broken with this "fix".

To sum up: We have to move the base class to the sample and archetype and give freedom to the projects using OASP. We have interface to standardize entities (with `getId()` and `getModificationCounter()`, etc.) what is entirely sufficient. Then projects also have the freedom to choose between JPA annotations on private fields or getter methods.
If actually hibernate would correctly support overriding JPA annotations of getter methods all would not be such a problem, but hibernate is not able to properly support this. Therefore I see no other way...

**Comment by hohwille** [comment]
BTW: Can anybody explain how to use flyway properly at all without sequences for IDs? IMHO that can not lead to a professional approach. Simply do not use MySQL and you are fine ;)

**Comment by hohwille** [comment]
BTW: `AUTO` also seems not to work for MSSQL:
http://stackoverflow.com/questions/25047226/jpa-generationtype-auto-not-considering-column-with-auto-increment

**Comment by hohwille** [comment]
FYI: All you need to do is ignore `AbstractPersistenceEntity` and define your own:
https://github.com/oasp/oasp4j/blob/develop/modules/jpa/src/main/java/io/oasp/module/jpa/dataaccess/api/AbstractPersistenceEntity.java
Fortunately we did that in my project so I can be relaxed :)
Anyway this just underlines my proposal. We need to move (copy) this class from the module to the sample (and archetype). However, we should keep it deprecated in its original place for ~2 releases.

**Comment by maybeec** [comment]
Sry to reopent this ticket so late. I am currently in charge for Hibernate issues in our project regarding the ID generation done by Hibernate and the problems of upgrading to Hibernate 5.
I had done a lot research in this direction and debugged Hibernate with different configurations and versions to be really sure what is going on. Currently the oasp4j is using Hibernate 4.x.
Hibernate 4.x uses deprecated ID Generators by default. The better ones have been introduced in Hibernate 3.2.3 but have not made it into the default configuration for Hibernate 4: http://in.relation.to/2007/04/10/new-323-hibernate-identifier-generators/
However, Hibernate 5 has switched the default of ID generator use to the new ones: https://github.com/hibernate/hibernate-orm/blob/5.0/migration-guide.adoc#changed-setting-defaults

The good about the new generators is, that it introduces a SequenceStyleGenerator, which is much more portable and will not depend on the support of sequences of the underlying database. It would rather simulate sequences by possibly create sequence tables in databases like mysql to simulate the same behavior. I would recommend to go into this direction and use the new sequencestylegenerators of hibernate to also prevent upgrade issues to hibernate 5.

There are many pitfalls you have to understand when configuring Hibernate, i.e., JPA Annotations are possibly interpreted differently than you might think of. As an example use GenerationType.SEQUENCE, which causes Hibernate to use a Sequence-HiLo strategy for ID-Generation rather than using a real sequence. In addition, GenerationType.AUTO will tread ID generation differently for each underlying database.

Furthermore, when considering GenerationType.SEQUENCE, there is a hidden Gem of a sequence generator optimization strategy called POOLED-LO, whereas the default strategy is POOLED. In contrast to POOLED, POOLED-LO stores the lo-value in the database rather than the hi-value, which makes interoperability of multiple systems on the same database possible: https://vladmihalcea.com/2014/07/21/hibernate-hidden-gem-the-pooled-lo-optimizer/

My proposal: Use GenerationType.SEQUENCE with POOLED-LO optimization strategy and adjust the performance needs by the allocationSize (JPA) resp. increment_size (Hibernate equivalent) parameter. Please be aware of the fact, that you have to adjust the allocationSize with the INCREMENT BY definition of your database sequence anyhow to not run into problems.

**Comment by hohwille** [comment]
See:
https://github.com/oasp/oasp4j/pull/436

Unfortunately this PR was not connected. We have to follow our own conventions:
https://github.com/oasp/oasp4j/wiki/oasp-code-contributions

**Comment by hohwille** [comment]
We deprecated `AbstractPersistenceEntity` but we should also add an `@deprecated` JavaDoc tag and explain the reason for the deprecation and what to do.
```
@deprecated will be removed in a future release. In order to give OASP users more flexibility we want to stop providing an JPA entity base class in this library. Instead we provide it with our application template (oasp4j-template-server) so you can take over control of JPA annotations. If you already started with OASP in an earlier version you can simply update `ApplicationPersistenceEntity` from our current sample on github to get rid of the dependency to this class.
```

**Comment by jhcore** [comment]
Done https://github.com/oasp/oasp4j/pull/438. Please @hohwille, can you confirm it?

**Comment by hohwille** [comment]
Perfect. :+1: Thanks.
Merged.
IMHO we are done with this issue.
We will close - any concerns?

**Comment by jhcore** [comment]
@hohwille yes, it's all ok. We can close this issue
